File: backend/app/api/routes/players.py
import csv
import io

from fastapi import APIRouter, Depends, File, HTTPException, UploadFile
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db, require_role
from app.models.player import Player
from app.models.user import User
from app.schemas.player import PlayerCreate, PlayerOut

router = APIRouter(prefix="/players", tags=["players"])

@router.post("", response_model=PlayerOut)
def create_player(
    payload: PlayerCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_role("coach")),
):
    player = Player(club_id=current_user.club_id, **payload.model_dump())
    db.add(player)
    db.commit()
    db.refresh(player)
    return player


@router.post("/import")
def import_players_csv(
    file: UploadFile = File(...),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_role("coach")),
):
    """
    Import players from CSV.

    Expected headers: first_name,last_name,jersey_number,position
    Upsert strategy (MVP): match by jersey_number within the club.
    """
    raw = file.file.read()
    try:
        text = raw.decode("utf-8-sig")
    except UnicodeDecodeError:
        raise HTTPException(status_code=400, detail="CSV must be UTF-8 encoded")

    reader = csv.DictReader(io.StringIO(text))
    required = {"first_name", "last_name", "jersey_number"}
    if not reader.fieldnames or not required.issubset(set(reader.fieldnames)):
        raise HTTPException(
            status_code=400,
            detail=f"Missing required headers. Required: {sorted(required)}",
        )

    created = 0
    updated = 0
    errors: list[dict] = []

    for i, row in enumerate(reader, start=2):  # header is line 1
        try:
            first = (row.get("first_name") or "").strip()
            last = (row.get("last_name") or "").strip()
            jersey_raw = (row.get("jersey_number") or "").strip()
            pos = (row.get("position") or "").strip() or None

            if not first or not last or not jersey_raw:
                raise ValueError("first_name, last_name, jersey_number are required")

            jersey = int(jersey_raw)
            if jersey <= 0 or jersey > 99:
                raise ValueError("jersey_number must be in 1..99")

            existing = (
                db.query(Player)
                .filter(Player.club_id == current_user.club_id, Player.jersey_number == jersey)
                .first()
            )
            if existing:
                existing.first_name = first
                existing.last_name = last
                existing.position = pos
                updated += 1
            else:
                db.add(
                    Player(
                        club_id=current_user.club_id,
                        first_name=first,
                        last_name=last,
                        jersey_number=jersey,
                        position=pos,
                    )
                )
                created += 1
        except Exception as e:  # noqa: BLE001 (CSV row-level error aggregation)
            errors.append({"line": i, "error": str(e), "row": row})

    db.commit()
    return {"ok": True, "created": created, "updated": updated, "errors": errors}

@router.get("", response_model=list[PlayerOut])
def list_players(db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    return (
        db.query(Player)
        .filter(Player.club_id == current_user.club_id)
        .order_by(Player.jersey_number)
        .all()
    )

@router.delete("/{player_id}")
def delete_player(
    player_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_role("coach")),
):
    p = (
        db.query(Player)
        .filter(Player.id == player_id, Player.club_id == current_user.club_id)
        .first()
    )
    if not p:
        raise HTTPException(status_code=404, detail="Player not found")
    db.delete(p)
    db.commit()
    return {"ok": True}

File: frontend/src/api/players.ts
import { api } from "./client";

export type Player = {
  id: number;
  first_name: string;
  last_name: string;
  jersey_number: number;
  position?: string | null;
};

export type ImportPlayersResult = {
  ok: boolean;
  created: number;
  updated: number;
  assigned?: number;
  season_id?: number;
  errors: Array<{ line: number; error: string; row: Record<string, unknown> }>;
};

export async function listPlayers(seasonId?: number) {
  const params = seasonId != null ? { season_id: seasonId } : undefined;
  const res = await api.get<Player[]>("/players", { params });
  return res.data;
}

export async function createPlayer(payload: {
  first_name: string;
  last_name: string;
  jersey_number: number;
  position?: string;
  season_id?: number;
}) {
  const res = await api.post<Player>("/players", payload);
  return res.data;
}

export async function deletePlayer(playerId: number) {
  const res = await api.delete(`/players/${playerId}`);
  return res.data;
}

export async function importPlayersCsv(file: File, seasonId?: number) {
  const form = new FormData();
  form.append("file", file);
  if (seasonId != null) {
    form.append("season_id", String(seasonId));
  }
  const res = await api.post("/players/import", form, {
    headers: { "Content-Type": "multipart/form-data" },
  });
  return res.data as ImportPlayersResult;
}

export async function exportPlayersCsv() {
  const res = await api.get("/export/players.csv", { responseType: "blob" });
  return res.data as Blob;
}